plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -256,534 +254,666 @@ struct block_time_tracker {
+
+   // path to write the snapshots to
+   std::filesystem::path _snapshots_dir;
+
+   // async snapshot scheduler
+   snapshot_scheduler _snapshot_scheduler;
+
+   std::function<void(producer_plugin::produced_block_metrics)> _update_produced_block_metrics;
+   std::function<void(producer_plugin::incoming_block_metrics)> _update_incoming_block_metrics;
+
+   // ro for read-only
+   struct ro_trx_t {
+      transaction_metadata_ptr trx;
+      next_func_t              next;
+   };
+   // The queue storing previously exhausted read-only transactions to be re-executed by read-only threads
+   // thread-safe
+   class ro_trx_queue_t {
+   public:
+      void push_front(ro_trx_t&& t) {
+         std::lock_guard g(mtx);
+         queue.push_front(std::move(t));
       }
 
-      void on_block_header( const block_state_ptr& bsp ) {
-         consider_new_watermark( bsp->header.producer, bsp->block_num, bsp->block->timestamp );
+      bool empty() const {
+         std::lock_guard g(mtx);
+         return queue.empty();
       }
 
-      void on_irreversible_block( const signed_block_ptr& lib ) {
-         const chain::controller& chain = chain_plug->chain();
-         EOS_ASSERT(chain.is_write_window(), producer_exception, "write window is expected for on_irreversible_block signal");
-         _irreversible_block_time = lib->timestamp.to_time_point();
-         _snapshot_scheduler.on_irreversible_block(lib, chain);
+      bool pop_front(ro_trx_t& t) {
+         std::unique_lock g(mtx);
+         if (queue.empty())
+            return false;
+         t = queue.front();
+         queue.pop_front();
+         return true;
       }
 
-      void abort_block() {
-         auto& chain = chain_plug->chain();
+   private:
+      mutable std::mutex mtx;
+      deque<ro_trx_t>    queue; // boost deque which is faster than std::deque
+   };
 
-         if( chain.is_building_block() ) {
-            _time_tracker.report( _idle_trx_time, chain.pending_block_num() );
-         }
-         _unapplied_transactions.add_aborted( chain.abort_block() );
-         _idle_trx_time = fc::time_point::now();
+   uint32_t _ro_thread_pool_size{0};
+   // Due to uncertainty to get total virtual memory size on a 5-level paging system for eos-vm-oc and
+   // possible memory exhuastion for large number of contract usage for non-eos-vm-oc, set a hard limit
+   static constexpr uint32_t         _ro_max_threads_allowed{8};
+   named_thread_pool<struct read>    _ro_thread_pool;
+   fc::microseconds                  _ro_write_window_time_us{200000};
+   fc::microseconds                  _ro_read_window_time_us{60000};
+   static constexpr fc::microseconds _ro_read_window_minimum_time_us{10000};
+   fc::microseconds                  _ro_read_window_effective_time_us{0}; // calculated during option initialization
+   std::atomic<int64_t>
+      _ro_all_threads_exec_time_us; // total time spent by all threads executing transactions. use atomic for simplicity and performance
```

Comment:
Hmm, don't like that it added a return because the comment was too long.

Reply:
fixed